accepted/0642-vendoring.md

```diff
@@ -0,0 +1,43 @@
+# Dependency vendoring
+
+## Summary
+
+Provide a mechanism to vendor all dependencies for a project in a more reproducible form than adding `node_modules` to version control.
+
+## Motivation
+
+A `node_modules` directory can differ between operating systems(/npm versions?), which presents pitfalls for projects that want to ensure full reproducibility of a given dependency tree, such as [Nixpkgs](https://github.com/NixOS/nixpkgs).
+
+## Detailed Explanation
+
+This would add a new command to the npm CLI, `npm vendor`. This command results in a reproducible {archive, directory} that contains all dependencies of the given project. Additionally, a configuration value will be added. When `npm install` and `npm ci` are ran with this value set to a valid path, they will pull from the vendored copies of dependencies.
+
+## Rationale and Alternatives
+
+To my knowledge, there are two alternatives:
+1. Ship `node_modules`. This would work a good percent of the time, assuming `npm ci --ignore-scripts` is used, and all packages are installed across all operating systems.
+2. Generate a cache, and point npm to that. Given that the cacache format hasn't changed in a number of years, nor has what npm stored in cache entry metadata, I'd say this is a pretty safe bet. However, I feel uneasy about relying on it, as it could break at any time.
```

Comment:
the third option is the best practice for a long time - use a lockfile, and an internal registry, and any resulting differences between systems of `node_modules` are intentional and necessary.

Reply:
I understand that there are differences between different `node_modules`, but having to run an internal registry certainly begs for a more lightweight solution, IMHO. (Why use an internal registry over the public one, assuming all of your dependencies are there, if you have a lockfile?)

Reply:
Because deploying from the internet is hugely unsafe and exposes you to DNS/man-in-the-middle attacks?

Any discussion about reproducibility rests on first having no uncontrolled sources of content.

Reply:
How is this the case with a known-good lockfile, which contains hashes for all dependencies, though?

Reply:
That's fair, but you still don't want the contents denied to you entirely - that breaks your build too.